tests/src/Kernel/Access/GroupLevelAccessTest.php

```diff
@@ -0,0 +1,450 @@
+    $this->assertFalse($this->ogAccess->userAccess($this->group, 'some_perm', $this->ownerUser)->isAllowed());
+  }
+
+  /**
+   * Sets up a matrix of users and roles with arbitrary permissions.
+   *
+   * @return array[]
+   *   A tuple containing the created test roles and users.
+   */
+  protected function setupUserAccessArbitraryPermissions() {
+    $roles = [];
+    $users = [];
+
+    // Create another group to test per group/per account permission caching.
+    // This is a group of the same entity type and bundle.
+    $alternate_group = EntityTest::create([
+      'type' => $this->groupBundle,
+      'name' => $this->randomString(),
+      'user_id' => $this->ownerUser->id(),
+    ]);
+    $alternate_group->save();
+
+    // Create a role with an arbitrary permission to test with.
+    $role = OgRole::create();
+    $role
+      ->setName($this->randomMachineName())
+      ->setLabel($this->randomString())
+      ->setGroupType($this->group->getEntityTypeId())
+      ->setGroupBundle($this->groupBundle)
+      // Associate an arbitrary permission with the role.
+      ->grantPermission('some_perm')
+      ->save();
+    $roles['arbitrary_permission'] = $role;
+
+    // Create a role with an arbitrary permission which will only be granted to
+    // a member of the second group.
+    $alternate_role = OgRole::create();
+    $alternate_role
+      ->setName($this->randomMachineName())
+      ->setLabel($this->randomString())
+      ->setGroupType($alternate_group->getEntityTypeId())
+      ->setGroupBundle($alternate_group->bundle())
+      // Associate an arbitrary permission with the role.
+      ->grantPermission('some_perm_2')
+      ->save();
+    $roles['alternate'] = $alternate_role;
+
+    // Create a user which is a member of both test groups and has an arbitrary
+    // permission in both. This allows us to test that permissions do not leak
+    // between different groups.
+    $user = User::create(['name' => $this->randomString()]);
+    $user->save();
+
+    /** @var \Drupal\og\OgMembershipInterface $membership */
+    $membership = Og::createMembership($this->group, $user);
+    $membership
+      ->addRole($role)
+      ->save();
+
+    $membership = Og::createMembership($alternate_group, $user);
+    $membership
+      ->addRole($alternate_role)
+      ->save();
+
+    $users['has_permission_in_both_groups'] = $user;
+
+    // Create a user which is a member and has a role without any permissions.
+    $user = User::create(['name' => $this->randomString()]);
+    $user->save();
+
+    $role_without_permissions = OgRole::create();
+    $role_without_permissions
+      ->setName($this->randomMachineName())
+      ->setLabel($this->randomString())
+      ->setGroupType($this->group->getEntityTypeId())
+      ->setGroupBundle($this->groupBundle)
+      ->save();
+
+    $membership = Og::createMembership($this->group, $user);
+    $membership
+      ->addRole($role_without_permissions)
+      ->save();
```

Comment:
I'd create a helper function to group this:

```
  function createTestUser($role) {
    // Create a user which is a member and has a role without any permissions.
    $user = User::create(['name' => $this->randomString()]);
    $user->save();

    $membership = Og::createMembership($this->group, $user);
    $membership
      ->addRole($role)
      ->save();
    
    return $user;
  }

```

Reply:
Great idea, this will make the setup more compact and easy to follow. I'll also make a helper method to create OG roles.